Fix issue #60 : Wiki page explaining commonly encountered Server Errors and how to prevent them (#242)
oppia/oppia-web-developer-docs: Server-errors-and-solutions.md

> This document talks about different types of server problems that have been fixed before, and it suggests ways to stop those problems from happening again. The types of problems include issues with data, problems when lots of people are using the server, situations that don't happen very often, making sure the server works with old internet browsers, mistakes in the way the server thinks, problems when the internet is slow, and errors that don't need to be written down. For each type of problem, it gives an example and tips on how to stop it from happening.

Reviewer: document --> page

talks about --> discusses

I don't understand parts of the second sentence ("errors that don't need to be written down", "mistakes in the way the server thinks"). I suggest you just get rid of that sentence altogether.

Author: I removed those 2 sentences for now. Do you want me to delete the whole sentence starting from "The type of problems"
till "written down"?

Reviewer: Yes, please do that.

Author: done

# Categories

* [Scalability issue](#scalability-issue)
* [Corner cases](#corner-cases)
* [Supporting old browsers](#supporting-old-browsers)
* [Logic errors](#logic-errors)
* [Data related issue](#data-related-issues)
* [Race conditions / issues on slow networks](#race-conditions)
* [Related to migration](#related-to-migrations)
* [Errors that didn't need to be logged](#errors-that-didn't-need-to-be-logged)

## Scalability Issue

> Issues that are related to features not scaling well on production. Example: (1) Exploration state data is too large(#12547) for some explorations and no further changes can be made to these explorations. (2) Suggestion models being fetched for review exceeds the query limit, so only an incomplete result is shown.

Reviewer: Link properly to the issue in question. Make sure the link works.

Put a space before the parentheses.

Author: Yes sir done

### Preventive action

* __Design stage__: During design doc reviews, one should ask questions about performance and scalability when dealing with large amounts of data – “how big do you expect X to get?”, “are there any limits to the size?”, “what happens when it reaches this limit and is it possible to handle it?”

> * __Manual testing__: Having a functionality to generate dummy data for testing locally might help.

Reviewer: a --> some

Author: Done

## Corner Cases

> User-facing issues where certain corner case scenarios are not considered (empty list, undefined object etc.) Example: Deleting a hint that is active in the editor view caused console errors. Although these are not often serious, these issues can be tricky to debug due to lack of reproduction steps. The majority of server errors fall under this category. Other e.g. error #7907, #6298, #6199, #12528, #8702, #8464, #7970, #7546, #7444, #7727 (usage of plurals in certain languages), #7342, #7331, #7047, #6757, #6523 (didn't account for changes to interaction type), #5457

Reviewer: Make sure your links work (they currently don't). Use Markdown format for links.

Also, you don't need so many examples, at most 5 will do.

Author: Now links are working

### Preventive action

> * __Static analysis / compilation stage__: Strongly typed frontend will help catch some of these issues.

Reviewer: Having strong typing in the frontend will...

Author: done

> * __Manual testing__: Developers generally test only the “happy” paths in their feature. We should encourage them to test edge cases. Adoption of template users in testing mentioned here should help uncover / test edge case scenarios.

Reviewer: The link for "here" is missing. Use https://docs.google.com/document/d/1ktMagEKvUA6A0UmM_2CoENZPelqmjLmRjV7tWeo6w54/edit#heading=h.y3evobat23c5

Author: Ah, my bad I forgot. Corrected it now

## Supporting old browsers

> Certain JS functionality may not be supported in older browsers. Example: String.prototype.match was used in the codebase but it lacked support in IE browsers. Fixes for these issues often involve adding a polyfill or swapping the relevant JS method with a more compatible one. Other e.g. errors #6909, #6174, #7926, #7794, #7784, #7205

Reviewer: Other e.g. errors --> Other examples of errors include: ...

Also please fix the links, here and elsewhere (I won't comment on this any more).

Author: Done

### Preventive action

* __Static analysis / compilation stage__: Fix a JS standard that we expect all clients to support and specify a “target version” in typescript so that the code will be compiled to that version / JS standard.

## Logic errors
Logical errors in the code. Example: “this” scope was referring to the global scope when it was supposed to refer to a local one (#7873). These are quite rare. Other e.g. errors #7873, #7872 (incorrect scope of this), #7927, #7460

### Preventive action

* __Code review__: These issues need to be flagged during code reviews. Often happens when the developer is refactoring existing code but hasn’t fully tested their changes.

## Data-related Issue

> Issues that resulted in corrupt or missing data on the server. Issues like this are quite hard to debug and involve complicated one-off jobs to be run. Example: Certain exploration stats models were missing on the production server, this resulted in few deferred tasks erroring often and piling up in the stackdriver logs. Fortunately, these issues are quite rare. Other e.g. errors #6159, #9632, #7440

Reviewer: this resulted --> which resulted

few --> a few

stackdriver --> StackDriver

Please correct "Other e.g. errors" as mentioned above, and apply similar corrections to other sections.

Author: Done

### Preventive action

* __Manual testing__: When testing, perform destructive operations on entities that have related supplementary entities to check if the supplementary objects reflect the correct state of the original ones.

> Testing with foreign / special characters (as mentioned under Template Users).

Reviewer: You should have 1., 2.  numbered bullet points here, similar to the original doc. Please preview your changes to make sure they are correct.

Author: Done

Reviewer 2: not done

Author: It didn't have any heading so I thought it will be another new line in manual testing para itself. 
okay Now I changed it

* __Code__: Audit jobs need to account for all instances or occurrences of the relevant item on the server. Sufficient written proof / explanation about the “thoroughness” of the audit can be added as a comment in the audit job code.

* __Review phase__: Reviewers should be encouraged to ask questions and ensure that the audit job is thorough.

## Race Conditions / Slow Network Issues

Some errors happen when interacting too fast with the application or when the network is too slow. Example: Clicking on the “content” tab in the translation editor before it has fully loaded (#6378). Other e.g. errors #6161, #13087, #8181, #8733, #8321, #7381, #7252, #7086, #7041, #6921, #6298, #6161   

### Preventive action

> * __Manual testing__: For code that has a lot of “async” functionality, the authors should be encouraged to try throttling their network and/or interacting with the UI before it has completely loaded. This could be a category of “template users” mentioned here.

Reviewer: The link under "here" is missing. See above.

Author: Done

* __Automated testing__: Configure certain automated tests to run under throttled network conditions. Might be worth considering the introduction of tests where the test interacts too quickly with elements on the page and expect nothing to happen (there should be no console errors).

## Related to a migration

One-off errors from migration of libraries or frameworks e.g. missing imports in the module files.

### Preventive action
* __Documentation__: Record common issues in the relevant migration page when they are identified.

## Errors that didn't need to be logged

> Sometimes there may be unexpected errors thrown that do not impact user actions. Such errors can be ignored on a case-by-case basis. Example: Navigating away from a page when there are pending requests or if the request timed out caused some console errors. Other e.g. errors: [$compile:tpload]" issue causing page load errors #5793, #6500, #8761, #7563 

Reviewer: Why is there a $compile:tpload here?

Author: This was mentioned it in the word file provided in issue thread. I removed it now

### Preventive action

> * __Code / review__: When adding logging statements or throwing errors that are not handled, always consider what scenarios and how often they will get called. If the frequency is high, discuss with the server error coordinator for the month to check if there are alternatives to achieve the same (e.g. it may be possible to add logs on stackdriver interface directly in a controlled way.)

Reviewer: what scenarios --> what scenarios these arise in

same --> desired logging behaviour

Author: done

This page discusses different types of server problems that have been fixed before, and it suggests ways to stop those problems from happening again. For each type of problem, it gives an example and tips on how to stop it from happening.

> # Categories

Reviewer 2: ```suggestion
## Table of contents
```

Author: done

* [Scalability issue](#scalability-issue)
* [Corner cases](#corner-cases)
* [Supporting old browsers](#supporting-old-browsers)
* [Logic errors](#logic-errors)
* [Data related issue](#data-related-issue)
* [Race Conditions / Slow Network Issues](#race-conditions--slow-network-issues)
* [Related to migration](#related-to-migration)
* [Errors that didn't need to be logged](#errors-that-didnt-need-to-be-logged)

> ## Scalability Issue

Reviewer 2: Headings should be sentence-case (i.e. "Issue" -> "issue"). Ditto elsewhere

Author: done

> Issues that are related to features not scaling well on production. Example: (1) Exploration state data is too large (#12547) for some explorations and no further changes can be made to these explorations. (2) Suggestion models being fetched for review exceeds the query limit, so only an incomplete result is shown.

Reviewer 2: As seanlip previously commented (https://github.com/oppia/oppia-web-developer-docs/pull/242#discussion_r1457818812), the issue link doesn't work. You can see this in the file preview here: https://github.com/oppia/oppia-web-developer-docs/blob/de2dfc2a44167ce568380c4d862021dfae0e6570/Server-errors-and-solutions.md. GitHub only auto-generates links for issues/PRs in issue, PR, and Discussion threads, not in general Markdown files or in the wiki.

In your video in https://github.com/oppia/oppia-web-developer-docs/pull/242#issuecomment-1901101129, you can see that the issue references aren't made into links.

Ditto elsewhere

Author: oh that issue. Now I get it I thought "scalability issues" link was not working. sorry
I linked issues with links now. I removed some issues whose links I was not able to find.

### Preventive action

> 1. __Design stage__: During design doc reviews, one should ask questions about performance and scalability when dealing with large amounts of data – “how big do you expect X to get?”, “are there any limits to the size?”, “what happens when it reaches this limit and is it possible to handle it?”

Reviewer 2: Likely because you copied text from the doc, you have curly quotes (`“`) which curl the wrong way. Please replace them with straight quotes (`"`), which is what usually gets generated when you type quotation marks. Programs like Google Docs and Word just automatically replace straight quotes with the curly ones.

Ditto elsewhere

Author: done

2. __Manual testing__: Having some functionality to generate dummy data for testing locally might help.

## Corner Cases

> User-facing issues where certain corner case scenarios are not considered (empty list, undefined object etc.) Example: Deleting a hint that is active in the editor view caused console errors. Although these are not often serious, these issues can be tricky to debug due to lack of reproduction steps. The majority of server errors fall under this category. Other examples of errors include: #7907, #7727 (usage of plurals in certain languages), #7342, #6523 (didn't account for changes to interaction type), #5457

Reviewer 2: ```suggestion
User-facing issues where certain corner case scenarios are not considered (e.g. empty list or undefined object). Example: Deleting a hint that is active in the editor view caused console errors. Although these are not often serious, these issues can be tricky to debug due to lack of reproduction steps. The majority of server errors fall under this category. Other examples of errors include: #7907, #7727 (usage of plurals in certain languages), #7342, #6523 (didn't account for changes to interaction type), #5457
```

Re-structure sentence to avoid ending with "etc." since ending with a period-terminated abbreviation creates ambiguity about whether the sentence has ended

Author: done

### Preventive action

1. __Static analysis / compilation stage__: Having strong typing in the frontend will help catch some of these issues.
2. __Manual testing__: Developers generally test only the “happy” paths in their feature. We should encourage them to test edge cases. Adoption of template users in testing mentioned [here](https://docs.google.com/document/d/1ktMagEKvUA6A0UmM_2CoENZPelqmjLmRjV7tWeo6w54/edit#heading=h.y3evobat23c5) should help uncover / test edge case scenarios.

## Supporting old browsers

> Certain JS functionality may not be supported in older browsers. Example: String.prototype.match was used in the codebase but it lacked support in IE browsers. Fixes for these issues often involve adding a polyfill or swapping the relevant JS method with a more compatible one. Other examples of errors include: #6909, #6174, #7926, #7794

Reviewer 2: ```suggestion
Certain JS functionality may not be supported in older browsers. Example: String.prototype.match was used in the codebase but it lacked support in IE browsers. Fixes for these issues often involve adding a polyfill or swapping the relevant JS method with a more compatible one. Other examples of errors include: #6909, #6174, #7926, #7794.
```

Ditto elsewhere for ending sentences with punctuation.

Author: done

### Preventive action

1. __Static analysis / compilation stage__: Fix a JS standard that we expect all clients to support and specify a “target version” in typescript so that the code will be compiled to that version / JS standard.

## Logic errors

> Logical errors in the code. Example: “this” scope was referring to the global scope when it was supposed to refer to a local one (#7873). These are quite rare. Other examples of errors include: #7873, #7872 (incorrect scope of this), #7927, #7460

Reviewer 2: ```suggestion
Logical errors in the code. Example: The “this” scope was referring to the global scope when it was supposed to refer to a local one (#7873). These are quite rare. Other examples of errors include: #7873, #7872 (incorrect scope of "this"), #7927, #7460
```

Author: done

### Preventive action

> 1. __Code review__: These issues need to be flagged during code reviews. Often happens when the developer is refactoring existing code but hasn’t fully tested their changes.

Reviewer 2: ```suggestion
1. __Code review__: These issues need to be flagged during code reviews. They often happen when a developer is refactoring existing code but hasn’t fully tested their changes.
```

Author: done

## Data-related Issue

> Issues that resulted in corrupt or missing data on the server. Issues like this are quite hard to debug and involve complicated one-off jobs to be run. Example: Certain exploration stats models were missing on the production server, which resulted in a few deferred tasks erroring often and piling up in the StackDriver logs. Fortunately, these issues are quite rare. Other examples of errors include: #6159, #9632, #7440

Reviewer 2: ```suggestion
Issues that result in corrupt or missing data on the server. Issues like this are quite hard to debug and involve running complicated one-off jobs. Example: Certain exploration stats models were missing on the production server, which resulted in a few deferred tasks erroring often and piling up in the StackDriver logs. Fortunately, these issues are quite rare. Other examples of errors include: #6159, #9632, #7440
```

Let's keep the issue descriptions in present tense. Please also proofread for unnecessarily convoluted language like "complicated one-off jobs to be run"

Author: done

### Preventive action

1. __Manual testing__: When testing, perform destructive operations on entities that have related supplementary entities to check if the supplementary objects reflect the correct state of the original ones.

> Testing with foreign / special characters (as mentioned under Template Users).

Reviewer 2: Not clear what "Template Users" is referring to. Do you need to link to a doc?

Author: yeah done

### Preventive action

1. __Manual testing__: For code that has a lot of “async” functionality, the authors should be encouraged to try throttling their network and/or interacting with the UI before it has completely loaded. This could be a category of “template users” mentioned [here](https://docs.google.com/document/d/1ktMagEKvUA6A0UmM_2CoENZPelqmjLmRjV7tWeo6w54/edit#heading=h.y3evobat23c5)

2. __Automated testing__: Configure certain automated tests to run under throttled network conditions. Might be worth considering the introduction of tests where the test interacts too quickly with elements on the page and expect nothing to happen (there should be no console errors).

## Related to migration

One-off errors from migration of libraries or frameworks e.g. missing imports in the module files.

### Preventive action

1. __Documentation__: Record common issues in the relevant migration page when they are identified.

## Errors that didn't need to be logged

> Sometimes there may be unexpected errors thrown that do not impact user actions. Such errors can be ignored on a case-by-case basis. Example: Navigating away from a page when there are pending requests or if the request timed out caused some console errors. Other examples of errors include: #5793, #6500, #8761, #7563 

Reviewer 2: ```suggestion

Sometimes there may be unexpected errors thrown that do not impact user actions. Such errors can be ignored on a case-by-case basis. Example: Navigating away from a page when there are pending requests or if the request timed out caused some console errors. Other examples of errors include: #5793, #6500, #8761, #7563 
```

Author: done

### Preventive action

> 1. __Code / review__: When adding logging statements or throwing errors that are not handled, always consider what scenarios these arise in and how often they will get called. If the frequency is high, discuss with the server error coordinator for the month to check if there are alternatives to achieve the desired logging behaviour (e.g. it may be possible to add logs on stackdriver interface directly in a controlled way.)

Reviewer 2: stackdriver -> StackDriver

Author: I perfectly remember making these changes before when seanlip told. Don't know why it didn't get updated.
I did it again.

2. __Manual testing__: Developers generally test only the "happy" paths in their feature. We should encourage them to test edge cases. Adoption of template users in testing mentioned [here](https://docs.google.com/document/d/1ktMagEKvUA6A0UmM_2CoENZPelqmjLmRjV7tWeo6w54/edit#heading=h.y3evobat23c5) should help uncover / test edge case scenarios.

## Supporting old browsers

Certain JS functionality may not be supported in older browsers. Example: String.prototype.match was used in the codebase but it lacked support in IE browsers. Fixes for these issues often involve adding a polyfill or swapping the relevant JS method with a more compatible one. Other examples of errors include: [#6909](https://github.com/oppia/oppia/pull/6909), [#6174](https://github.com/oppia/oppia/pull/6174), [#7926](https://github.com/oppia/oppia/pull/7926).

### Preventive action

1. __Static analysis / compilation stage__: Fix a JS standard that we expect all clients to support and specify a "target version" in typescript so that the code will be compiled to that version / JS standard.

## Logic errors

> Logical errors in the code. Example: The "this" scope was referring to the global scope when it was supposed to refer to a local one ([#7873](https://github.com/oppia/oppia/pull/7873)). These are quite rare. Other examples of errors include: [#7872](https://github.com/oppia/oppia/issues/7872)(incorrect scope of "this"), [#7927](https://github.com/oppia/oppia/pull/7927).

Reviewer 2: ```suggestion
Logical errors in the code. Example: The "this" scope was referring to the global scope when it was supposed to refer to a local one ([#7873](https://github.com/oppia/oppia/pull/7873)). These are quite rare. Other examples of errors include: [#7872](https://github.com/oppia/oppia/issues/7872) (incorrect scope of "this"), [#7927](https://github.com/oppia/oppia/pull/7927).
```

Please proofread the rendered version to catch typos like this

Author: but their shouldn't be a gap between issue no. and bracket right?  I mean in general we write it as for eg.
28(twenty eight). So should I change it?

Reviewer 2: There should be a gap between the issue number and parenthesis, and your example should be written as "28 (twenty eight)" with a space.

The only time I can think of where there wouldn't be a space before a parenthesis would be if you had a hierarchy of sections in a document. For example, if a question on an exam had a question number 1 with a subpart a, you could refer to the subpart as 1(a).

Author: done

### Preventive action

1. __Code review__: These issues need to be flagged during code reviews. They often happen when a developer is refactoring existing code but hasn’t fully tested their changes.

## Data-related issue

> Issues that result in corrupt or missing data on the server. Issues that result in corrupt or missing data on the server. Issues like this are quite hard to debug and involve running complicated one-off jobs. Example: Certain exploration stats models were missing on the production server, which resulted in a few deferred tasks erroring often and piling up in the StackDriver logs. Fortunately, these issues are quite rare. Other examples of errors include: [#6159](https://github.com/oppia/oppia/pull/6159), [#9632](https://github.com/oppia/oppia/pull/9632), [#7440](https://github.com/oppia/oppia/pull/7440).

Reviewer 2: ```suggestion
Issues that result in corrupt or missing data on the server. Issues like this are quite hard to debug and involve running complicated one-off jobs. Example: Certain exploration stats models were missing on the production server, which resulted in a few deferred tasks erroring often and piling up in the StackDriver logs. Fortunately, these issues are quite rare. Other examples of errors include: [#6159](https://github.com/oppia/oppia/pull/6159), [#9632](https://github.com/oppia/oppia/pull/9632), [#7440](https://github.com/oppia/oppia/pull/7440).
```

Author: oh no, done

### Preventive action

1. __Manual testing__: When testing, perform destructive operations on entities that have related supplementary entities to check if the supplementary objects reflect the correct state of the original ones.

> 2. Testing with foreign / special characters (as mentioned under [Template Users](https://docs.google.com/document/d/1FYShgUHI0GM2I6eWe5B78PAQLvPxZxMFVjTivVoBrTg/edit#)).

Reviewer 2: This doesn't match the [original doc](https://docs.google.com/document/d/1hIQIaFNVvyNiZBcCx0X7-_2lc0r1yvjxGfn4GJMjwEs/edit#heading=h.jlsplhp0gps7), and it's a little confusing. Performing destructive operations and using foreign/special characters are two types of manual testing that could help prevent this issue, so I think we actually want something like this:

> 1. __Manual testing__:
>    1. When testing, perform destructive operations on entities that have related supplementary entities to check if the supplementary objects reflect the correct state of the original ones.
>    2. Testing with foreign / special characters (as mentioned under [Template Users](https://docs.google.com/document/d/1FYShgUHI0GM2I6eWe5B78PAQLvPxZxMFVjTivVoBrTg/edit#)).

Author: done

### Preventive action

1. __Code review__: These issues need to be flagged during code reviews. They often happen when a developer is refactoring existing code but hasn’t fully tested their changes.

## Data-related issue

Issues that result in corrupt or missing data on the server. Issues like this are quite hard to debug and involve running complicated one-off jobs. Example: Certain exploration stats models were missing on the production server, which resulted in a few deferred tasks erroring often and piling up in the StackDriver logs. Fortunately, these issues are quite rare. Other examples of errors include: [#6159](https://github.com/oppia/oppia/pull/6159), [#9632](https://github.com/oppia/oppia/pull/9632), [#7440](https://github.com/oppia/oppia/pull/7440).

### Preventive action

1. __Manual testing__:  
   i\. When testing, perform destructive operations on entities that have related supplementary entities to check if the supplementary objects reflect the correct state of the original ones.

>    ii\. Testing with foreign / special characters (as mentioned under [Template Users](https://docs.google.com/document/d/1FYShgUHI0GM2I6eWe5B78PAQLvPxZxMFVjTivVoBrTg/edit#)).

Reviewer 2: You don't need to write out the `i` and `ii`. If you put in `1.` and `2.` just like any other numbered list, the rendered page will automatically transform to `i` and `ii`. As an additional benefit, the wrapped lines will be indented properly

Author: okay done
